type_decoders/README.md

```diff
@@ -0,0 +1,787 @@
+```
+
+`isObject()` accepts a `key: Decoder` argument object and returns a new decoder that will verify that an input is a non-null object, that the input has the keys specified in the argument object, and that the key values pass the relevant decoder provided by the argument object. On `DecoderSuccess`, the a new object is returned and the key-values of that object are provided by output from the argument decoders.
+
+### isExactObject()
+
+```ts
+interface IExactObjectDecoderOptions<T> {
+  msg?: DecoderErrorMsgArg;
+  keyMap?: { [P in keyof T]?: string | number };
+}
+
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): Decoder<T> ;
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> | PromiseDecoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): PromiseDecoder<T> ;
+```
+
+`isExactObject()` is the same as `isObject()`, except the input object cannot have any excess properties.
+
+### isDictionary()
+
+```ts
+interface IDictionaryDecoderOptions<V> {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isDictionary<R, V = unknown>(decoder: Decoder<R, V>, options?: IDictionaryDecoderOptions<V>): Decoder<R[], V>;
+function isDictionary<R, V = unknown>(decoder: PromiseDecoder<R, V>, options?: IDictionaryDecoderOptions<V>): PromiseDecoder<R[], V>;
+```
+
+`isDictionary()` receives a decoder argument and uses that decoder to process all values (regardless of key) of an input object.
+
+### isArray()
+
+```ts
+interface IArrayDecoderOptions<V> {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isArray<R = unknown, V = unknown>(options?: IArrayDecoderOptions<V>): Decoder<R[], V>;
+function isArray<R, V = unknown>(decoder: Decoder<R, V>, options?: IArrayDecoderOptions<V>): Decoder<R[], V>;
+function isArray<R, V = unknown>(decoder: PromiseDecoder<R, V>, options?: IArrayDecoderOptions<V>): PromiseDecoder<R[], V>;
+```
+
+`isArray()` can be used to make sure an input is an array. If an optional decoder argument is provided, that decoder will be used to process all of the input's elements.
+
+### isTuple()
+
+```ts
+interface ITupleDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isTuple<Tuple extends [unknown, ...unknown[]]>(decoders: { [I in keyof Tuple]: Decoder<Tuple[I]> }, options?: ITupleDecoderOptions): Decoder<Tuple>;
+function isTuple<Tuple extends [unknown, ...unknown[]]>(decoders: { [I in keyof Tuple]: Decoder<Tuple[I]> | PromiseDecoder<Tuple[I]> }, options?: ITupleDecoderOptions): PromiseDecoder<Tuple>;
+```
+
+`isTuple()` receives an array of decoders and creates a decoder which can be used to verify that an input is:
+
+1. An array of the same length as the decoder argument array.
+2. The first decoder argument will be used the process the first element of an input array.
+3. The second decoder argument will be used the process the second element of an input array.
+4. etc...
+
+### isLazy()
```

Comment:
When would this be used? Would it be when the `Decoder` tree is expensive to generate?

Reply:
Good question. I should make the description clearer by providing an example. It's for a recursive data structure. For example:

```ts
const treeDecoder = isObject({
  value: isString(),
  children: isArray(isLazy(() => treeDecoder))
})
```

This object would have the interface 

```ts
interface ITreeDecoder {
  value: string;
  children: ITreeDecoder[];
}
```

Reply:
It could be named `isRecursive`. I chose the name because its lazily evaluating the decoder, but on reflection `isRecursive` probably makes more sense.

Reply:
I ended up changing the name back to `isLazy()`. This is because while I imagine the primary use case will be decoding recursive data structures, what the decoder is actually *doing* is lazily evaluating a function each time `decode()` is called to retrieve the decoder it should use. I imagine `isLazy()` has other use cases beyond simply recursive data structures.

I added some more detail in the API explanation to describe the must common use case.